src/main/java/org/apache/flink/benchmark/functions/IntegerLongSource.java

```diff
@@ -41,12 +42,16 @@ public IntegerLongSource(int numberOfKeys, long numberOfElements) {
 
     @Override
     public void run(SourceContext<Record> ctx) throws Exception {
-        long counter = 0;
+        final int parallelism = getRuntimeContext().getNumberOfParallelSubtasks();
+        long counter = getRuntimeContext().getIndexOfThisSubtask();
 
         while (running && counter < numberOfElements) {
             synchronized (ctx.getCheckpointLock()) {
                 ctx.collectWithTimestamp(Record.of((int) (counter % numberOfKeys), counter), counter);
-                counter++;
+                counter += parallelism;
+                if ((counter / parallelism) % (1024 * 16) == 0) {
```

Comment:
Please add a comment which at least explains the reasoning behind these values.

Also, you could simplify the calculation to reduce the impact on the source's performance. I guess the least impact is by using a simple `int emitWatermarkCounter` that starts with the value you'd like for the repetition, decreases every round, and is reset once 0 is reached.